host/docs/shd_cal_tx_dc_offset.1
.TH "shd_cal_tx_dc_offset" "1" "3.7.0" "SHD" "User Commands"
.SH NAME
shd_cal_tx_dc_offset \- Generate TX DC Offset Calibration Table for a SHD Device
.SH DESCRIPTION
The Universal Software Radio Peripheral Hardware Drivers handle calibration
information. This application reduces TX DC offset within the specified frequency
range using transmit leakage into the receive path. Remove all external hardware
from the RF antenna ports before using this application. This can take up to
several minutes with default settings.
.LP
NOTE: This application can only be used with the WBX, SBX, XCVR2450, and RFX daughterboards.
.SH OPTIONS
.TP
\fB\-\-help\fR
help message
.TP
\fB\-\-verbose\fR
enable some verbose
.TP
\fB\-\-args\fR arg
device address args [default = ""]
.TP
\fB\-\-tx_wave_freq\fR arg (=507123)
Transmit wave frequency in Hz
.TP
\fB\-\-tx_wave_ampl\fR arg (=0.69999999999999996)
Transmit wave amplitude in counts
.TP
\fB\-\-rx_offset\fR arg (=934400)
RX LO offset from the TX LO in Hz
.TP
\fB\-\-freq_start\fR arg
Frequency start in Hz (will use beginning of frequency range if unspecified)
.TP
\fB\-\-freq_stop\fR arg
Frequency stop in Hz (will use end of frequency range if unspecified)
.TP
\fB\-\-freq_step\fR arg (=7300000)
Step size for LO sweep in Hz
.TP
\fB\-\-nsamps\fR arg (=10000)
Samples per data capture
.PP
.SH SEE ALSO
SHD documentation:
.B http://files.ettus.com/manual/
.LP
GR-SHD documentation:
.B http://gnuradio.org/doc/doxygen/page_shd.html
.LP
Other SHD programs:
.sp
shd_cal_rx_iq_balance(1) shd_cal_tx_iq_balance(1)
.SH AUTHOR
This manual page was written by Maitland Bottoms and Nicholas Corgan
for the Debian project (but may be used by others).
.SH COPYRIGHT
Copyright (c) 2012 Ettus Research LLC
.LP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
.LP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


host/utils/shd_cal_tx_dc_offset.cpp
//
// Copyright 2010,2012,2014 Ettus Research LLC
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

#include "smini_cal_utils.hpp"
#include <shd/utils/thread_priority.hpp>
#include <shd/utils/safe_main.hpp>
#include <shd/utils/paths.hpp>
#include <shd/utils/algorithm.hpp>
#include <shd/utils/msg.hpp>
#include <shd/smini/multi_smini.hpp>
#include <boost/program_options.hpp>
#include <boost/format.hpp>
#include <boost/thread/thread.hpp>
#include <boost/math/special_functions/round.hpp>
#include <iostream>
#include <complex>
#include <ctime>

namespace po = boost::program_options;

/***********************************************************************
 * Transmit thread
 **********************************************************************/
static void tx_thread(shd::smini::multi_smini::sptr smini, const double tx_wave_freq, const double tx_wave_ampl)
{
    shd::set_thread_priority_safe();

    // set max TX gain
    smini->set_tx_gain(smini->get_tx_gain_range().stop());

    //create a transmit streamer
    shd::stream_args_t stream_args("fc32"); //complex floats
    shd::tx_streamer::sptr tx_stream = smini->get_tx_stream(stream_args);

    //setup variables and allocate buffer
    shd::tx_metadata_t md;
    md.has_time_spec = false;
    std::vector<samp_type> buff(tx_stream->get_max_num_samps()*10);

    //values for the wave table lookup
    size_t index = 0;
    const double tx_rate = smini->get_tx_rate();
    const size_t step = boost::math::iround(wave_table_len * tx_wave_freq / tx_rate);
    wave_table table(tx_wave_ampl);

    //fill buff and send until interrupted
    while (not boost::this_thread::interruption_requested())
    {
        for (size_t i = 0; i < buff.size(); i++)
            buff[i] = table(index += step);
        tx_stream->send(&buff.front(), buff.size(), md);
    }

    //send a mini EOB packet
    md.end_of_burst = true;
    tx_stream->send("", 0, md);
}

/***********************************************************************
 * Tune RX and TX routine
 **********************************************************************/
static double tune_rx_and_tx(shd::smini::multi_smini::sptr smini, const double tx_lo_freq, const double rx_offset)
{
    //tune the transmitter with no cordic
    shd::tune_request_t tx_tune_req(tx_lo_freq);
    tx_tune_req.dsp_freq_policy = shd::tune_request_t::POLICY_MANUAL;
    tx_tune_req.dsp_freq = 0;
    smini->set_tx_freq(tx_tune_req);

    //tune the receiver
    double rx_freq = smini->get_tx_freq() - rx_offset;
    double min_fe_rx_freq = smini->get_fe_rx_freq_range().start();
    double max_fe_rx_freq = smini->get_fe_rx_freq_range().stop();
    shd::tune_request_t rx_tune_req(rx_freq);
    rx_tune_req.dsp_freq_policy = shd::tune_request_t::POLICY_MANUAL;
    rx_tune_req.dsp_freq = 0;
    if (rx_freq < min_fe_rx_freq)
        rx_tune_req.dsp_freq = rx_freq - min_fe_rx_freq;
    else if (rx_freq > max_fe_rx_freq)
        rx_tune_req.dsp_freq = rx_freq - max_fe_rx_freq;
    smini->set_rx_freq(rx_tune_req);

    //wait for the LOs to become locked
    boost::this_thread::sleep(boost::posix_time::milliseconds(50));
    boost::system_time start = boost::get_system_time();
    while (not smini->get_tx_sensor("lo_locked").to_bool() or not smini->get_rx_sensor("lo_locked").to_bool())
    {
        if (boost::get_system_time() > start + boost::posix_time::milliseconds(100))
            throw std::runtime_error("timed out waiting for TX and/or RX LO to lock");
    }

    return smini->get_tx_freq();
}

/***********************************************************************
 * Main
 **********************************************************************/
int SHD_SAFE_MAIN(int argc, char *argv[])
{
    std::string args, subdev, serial;
    double tx_wave_freq, tx_wave_ampl, rx_offset;
    double freq_start, freq_stop, freq_step;
    size_t nsamps;
    double precision;

    po::options_description desc("Allowed options");
    desc.add_options()
        ("help", "help message")
        ("verbose", "enable some verbose")
        ("args", po::value<std::string>(&args)->default_value(""), "device address args [default = \"\"]")
        ("subdev", po::value<std::string>(&subdev), "Subdevice specification (default: first subdevice, often 'A')")
        ("tx_wave_freq", po::value<double>(&tx_wave_freq)->default_value(507.123e3), "Transmit wave frequency in Hz")
        ("tx_wave_ampl", po::value<double>(&tx_wave_ampl)->default_value(0.7), "Transmit wave amplitude")
        ("rx_offset", po::value<double>(&rx_offset)->default_value(.9344e6), "RX LO offset from the TX LO in Hz")
        ("freq_start", po::value<double>(&freq_start), "Frequency start in Hz (do not specify for default)")
        ("freq_stop", po::value<double>(&freq_stop), "Frequency stop in Hz (do not specify for default)")
        ("freq_step", po::value<double>(&freq_step)->default_value(default_freq_step), "Step size for LO sweep in Hz")
        ("nsamps", po::value<size_t>(&nsamps), "Samples per data capture")
        ("precision", po::value<double>(&precision)->default_value(default_precision), "Correction precision (default=0.0001)")
    ;

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);
    po::notify(vm);

    //print the help message
    if (vm.count("help")){
        std::cout << boost::format("SMINI Generate TX DC Offset Calibration Table %s") % desc << std::endl;
        std::cout <<
            "This application measures leakage between RX and TX on a transceiver daughterboard to self-calibrate.\n"
            "Note: Not all daughterboards support this feature. Refer to the SHD manual for details.\n"
            << std::endl;
        return EXIT_FAILURE;
    }

    // Create a SMINI device
    shd::smini::multi_smini::sptr smini = setup_smini_for_cal(args, subdev, serial);

    if (not vm.count("nsamps"))
        nsamps = size_t(smini->get_rx_rate() / default_fft_bin_size);

    //create a receive streamer
    shd::stream_args_t stream_args("fc32"); //complex floats
    shd::rx_streamer::sptr rx_stream = smini->get_rx_stream(stream_args);

    //create a transmitter thread
    boost::thread_group threads;
    threads.create_thread(boost::bind(&tx_thread, smini, tx_wave_freq, tx_wave_ampl));

    //re-usable buffer for samples
    std::vector<samp_type> buff;

    //store the results here
    std::vector<result_t> results;

    if (not vm.count("freq_start")) freq_start = smini->get_fe_tx_freq_range().start();
    if (not vm.count("freq_stop")) freq_stop = smini->get_fe_tx_freq_range().stop();

    //check start and stop frequencies
    if (freq_start < smini->get_fe_tx_freq_range().start())
    {
        std::cerr << "freq_start must be " << smini->get_fe_tx_freq_range().start() << " or greater for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }
    if (freq_stop > smini->get_fe_tx_freq_range().stop())
    {
        std::cerr << "freq_stop must be " << smini->get_fe_tx_freq_range().stop() << " or less for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }

    //check rx_offset
    double min_rx_offset = smini->get_rx_freq_range().start() - smini->get_fe_tx_freq_range().start();
    double max_rx_offset = smini->get_rx_freq_range().stop() - smini->get_fe_tx_freq_range().stop();
    if (rx_offset < min_rx_offset or rx_offset > max_rx_offset)
    {
        std::cerr << "rx_offset must be between " << min_rx_offset << " and "
            << max_rx_offset << " for this daughter board" << std::endl;
        return EXIT_FAILURE;
    }

    SHD_MSG(status) << boost::format("Calibration frequency range: %d MHz -> %d MHz") % (freq_start/1e6) % (freq_stop/1e6) << std::endl;

    //set RX gain
    smini->set_rx_gain(0);

    for (double tx_lo_i = freq_start; tx_lo_i <= freq_stop; tx_lo_i += freq_step)
    {
        const double tx_lo = tune_rx_and_tx(smini, tx_lo_i, rx_offset);

        //frequency constants for this tune event
        const double actual_rx_rate = smini->get_rx_rate();
        const double actual_tx_freq = smini->get_tx_freq();
        const double actual_rx_freq = smini->get_rx_freq();
        const double bb_dc_freq = actual_tx_freq - actual_rx_freq;

        //reset TX DC offset
        smini->set_tx_dc_offset(std::complex<double>(0, 0));

        //capture initial uncorrected value
        capture_samples(smini, rx_stream, buff, nsamps);
        const double initial_dc_dbrms = compute_tone_dbrms(buff, bb_dc_freq/actual_rx_rate);

        //bounds and results from searching
        double i_corr_start = -1.0;
        double i_corr_stop = 1.0;
        double i_corr_step = (i_corr_stop - i_corr_start)/(num_search_steps+1);
        double q_corr_start = -1.0;
        double q_corr_stop = 1.0;
        double q_corr_step= (q_corr_stop - q_corr_start)/(num_search_steps+1);
        double best_dc_dbrms = initial_dc_dbrms;
        double best_i_corr = 0;
        double best_q_corr = 0;
        while (i_corr_step >= precision or q_corr_step >= precision)
        {
            for (double i_corr = i_corr_start + i_corr_step; i_corr <= i_corr_stop - i_corr_step; i_corr += i_corr_step)
            {
                for (double q_corr = q_corr_start + q_corr_step; q_corr <= q_corr_stop - q_corr_step; q_corr += q_corr_step)
                {
                    const std::complex<double> correction(i_corr, q_corr);
                    smini->set_tx_dc_offset(correction);

                    //receive some samples
                    capture_samples(smini, rx_stream, buff, nsamps);
                    const double dc_dbrms = compute_tone_dbrms(buff, bb_dc_freq/actual_rx_rate);

                    if (dc_dbrms < best_dc_dbrms)
                    {
                        best_dc_dbrms = dc_dbrms;
                        best_i_corr = i_corr;
                        best_q_corr = q_corr;
                    }
                }
            }

            i_corr_start = best_i_corr - i_corr_step;
            i_corr_stop = best_i_corr + i_corr_step;
            i_corr_step = (i_corr_stop - i_corr_start)/(num_search_steps+1);
            q_corr_start = best_q_corr - q_corr_step;
            q_corr_stop = best_q_corr + q_corr_step;
            q_corr_step = (q_corr_stop - q_corr_start)/(num_search_steps+1);
        }

        if (best_dc_dbrms < initial_dc_dbrms)   //keep result
        {
            result_t result;
            result.freq = tx_lo;
            result.real_corr = best_i_corr;
            result.imag_corr = best_q_corr;
            result.best = best_dc_dbrms;
            result.delta = initial_dc_dbrms - best_dc_dbrms;
            results.push_back(result);
            if (vm.count("verbose"))
                std::cout << boost::format("TX DC: %f MHz: lowest offset %f dB, corrected %f dB") % (tx_lo/1e6) % result.best % result.delta << std::endl;
            else
                std::cout << "." << std::flush;
        }

    }

    std::cout << std::endl;

    //stop the transmitter
    threads.interrupt_all();
    boost::this_thread::sleep(boost::posix_time::milliseconds(500));    //wait for threads to finish
    threads.join_all();

    store_results(results, "TX", "tx", "dc", serial);

    return EXIT_SUCCESS;
}
